homeassistant/auth/session.py

```diff
@@ -0,0 +1,197 @@
+        if refresh_token_id in self._refresh_token_revoce_callbacks:
+            return
+
+        @callback
+        def async_invalidate_auth_sessions() -> None:
+            """Invalidate all sessions for a refresh token."""
+            self._strict_connection_sessions = {
+                session_id: token_id
+                for session_id, token_id in self._strict_connection_sessions.items()
+                if token_id != refresh_token_id
+            }
+            self._async_schedule_save()
+
+        self._refresh_token_revoce_callbacks[refresh_token_id] = (
+            self._auth.async_register_revoke_token_callback(
+                refresh_token_id, async_invalidate_auth_sessions
+            )
+        )
+
+    async def async_create_session(
+        self,
+        request: Request,
+        refresh_token: RefreshToken,
+    ) -> None:
+        """Create new session for given refresh token."""
+        if not self._auth.async_get_refresh_token(refresh_token.id):
+            return
+
+        self._strict_connection_sessions = {
+            session_id: token_id
+            for session_id, token_id in self._strict_connection_sessions.items()
+            if token_id != refresh_token.id
+        }
+
+        self._async_register_revoke_token_callback(refresh_token.id)
+        session_id = await self._async_create_new_session(request)
+        self._strict_connection_sessions[session_id] = refresh_token.id
+        self._async_schedule_save()
+
+    async def async_create_temp_unauthorized_session(self, request: Request) -> None:
+        """Create a temporary unauthorized session."""
+        session_id = await self._async_create_new_session(
+            request, max_age=int(TEMP_TIMEOUT_SECONDS)
+        )
+
+        def remove(_: datetime) -> None:
+            self._temp_sessions.pop(session_id, None)
+
+        self._temp_sessions[session_id] = StrictConnectionTempSessionData(
+            async_call_later(self._hass, TEMP_TIMEOUT_SECONDS, remove)
+        )
+
+    async def _async_create_new_session(
+        self,
+        request: Request,
+        *,
+        max_age: int | None = None,
+    ) -> str:
+        session_id = secrets.token_hex(64)
+
+        session = await new_session(request)
+        session[SESSION_ID] = session_id
+        if max_age is not None:
+            session.max_age = max_age
+        return session_id
+
+    @callback
+    def _async_schedule_save(self, delay: float = 1) -> None:
+        """Save sessions."""
+        self._store.async_delay_save(self._data_to_save, delay)
+
+    @callback
+    def _data_to_save(self) -> StoreData:
+        """Return the data to store."""
+        return StoreData(
+            unauthorized_sessions=self._strict_connection_sessions,
+            key=self.key,
+        )
+
+    async def async_setup(self) -> None:
+        """Set up session manager."""
+        data = await self._store.async_load()
+        if data is None or not isinstance(data, dict):
```

Comment:
What other types than None and the generic type can data be?

Reply:
None and StoreData (TypedDict) are the only types. I copied this check from `auth_store.py`

Reply:
Ok. Maybe we don't need the check for dict then?